src/games/strategy/triplea/ai/proAI/data/ProMoveOptions.java

```diff
@@ -0,0 +1,116 @@
+package games.strategy.triplea.ai.proAI.data;
+
+import games.strategy.engine.data.PlayerID;
+import games.strategy.engine.data.Territory;
+import games.strategy.engine.data.Unit;
+import games.strategy.triplea.ai.proAI.util.ProBattleUtils;
+import games.strategy.triplea.ai.proAI.util.ProUtils;
+import games.strategy.triplea.delegate.Matches;
+import games.strategy.util.Match;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class ProMoveOptions {
+
+  private final ProUtils utils;
+  private final ProBattleUtils battleUtils;
+  private final Map<Territory, ProTerritory> maxMoveMap;
+  private final Map<Territory, List<ProTerritory>> moveMaps;
+
+  public ProMoveOptions(final ProUtils utils, final ProBattleUtils battleUtils) {
+    this.utils = utils;
+    this.battleUtils = battleUtils;
+    maxMoveMap = new HashMap<Territory, ProTerritory>();
+    moveMaps = new HashMap<Territory, List<ProTerritory>>();
+  }
+
+  public ProMoveOptions(final ProUtils utils, final ProBattleUtils battleUtils,
+      final List<Map<Territory, ProTerritory>> moveMapList, final PlayerID player, final boolean isAttacker) {
+    this(utils, battleUtils);
+    populateMaxMoveMap(moveMapList, player, isAttacker);
+    populateMoveMaps(moveMapList);
+  }
+
+  public ProTerritory getMax(final Territory t) {
+    return maxMoveMap.get(t);
+  }
+
+  public List<ProTerritory> getAll(final Territory t) {
+    return moveMaps.get(t);
+  }
+
+  @Override
+  public String toString() {
+    return maxMoveMap.toString();
+  }
+
+  private void populateMaxMoveMap(final List<Map<Territory, ProTerritory>> moveMaps, final PlayerID player,
+      final boolean isAttacker) {
+
+    // Get players in turn order
+    final List<PlayerID> players = utils.getOtherPlayersInTurnOrder(player);
+
+    for (final Map<Territory, ProTerritory> moveMap : moveMaps) {
+      for (final Territory t : moveMap.keySet()) {
+
+        // Get current player
+        PlayerID movePlayer = null;
+        final Set<Unit> currentUnits = new HashSet<Unit>(moveMap.get(t).getMaxUnits());
+        currentUnits.addAll(moveMap.get(t).getMaxAmphibUnits());
+        if (!currentUnits.isEmpty()) {
+          movePlayer = currentUnits.iterator().next().getOwner();
+        } else {
+          continue;
+        }
+
+        // Check if mover's turn comes before territory owner's
+        if (!utils.isPlayersTurnFirst(players, movePlayer, t.getOwner())) {
```

Comment:
Functional question - is there missing logic here? Does this if check prevent consideration of allied reinforcements coming to block a can-opener. For example, in big world, Italy can block a UK can-opener that is to support Russia against germany.